### ISSUE TITLE
elm-pages dev stops working when I remove an if expression

### ISSUE THREAD

**Issue by MartinSStewart** [dillonkearns/elm-pages#491]
Steps to reproduce:
1. Check out this branch https://github.com/MartinSStewart/martins-homepage/tree/weird-bug
2. Start `npx elm-pages dev`
3. Open `localhost:1234`
4. You should get this error in the console ![image](https://github.com/user-attachments/assets/7cd99ee6-14a9-4b62-a60a-bbf15e7cdb14)

Note that this issue goes away if you go to the previous commit on the branch.

I'm on Windows and I can trigger this error both on Lamdera 1.3 and 1.2.2

**Comment by MartinSStewart** [comment]
I've narrowed it down to this
```elm
view : App Data ActionData RouteParams -> Shared.Model -> Model -> View (PagesMsg Msg)
view app _ model =
    { title = ""
    , body = Ui.el [ Ui.id "Elm" ] Ui.none
    }
```
I think what is happening `Html.Attribute.id "Elm"` is causing some name collision.

**Comment by MartinSStewart** [comment]
I've converted Index.elm into a SSCCE for this issue

**Comment by dillonkearns** [comment]
Thanks for the issue and SSCCE! I was trying to look through the codebase for anything that references an "elm" or "Elm" HTML id, but couldn't find it 🤔 

I also notice that the elm-pages docs site, for example, doesn't have any elements with the ID "elm" or "Elm".

![elm-pages - a statically typed site generator 2024-10-07 at 2 28 13 PM](https://github.com/user-attachments/assets/b5c95043-64bc-4d10-9cda-7420bf47cd27)

So I'm pretty confused at why this would cause a problem.

**Comment by MartinSStewart** [comment]
The module name usually is called Elm right? Maybe that can somehow have a name collision with DOM ids?

**Comment by lydell** [comment]
I ran your SSCCE, and also ran into an error, although not the same as yours.

```
Uncaught TypeError: cannot use 'in' operator to search for "init" in ""
    findPublicModules http://localhost:1234/elm.js:25173
    _elm_hot_loader_init http://localhost:1234/elm.js:25559
    <anonymous> http://localhost:1234/elm.js:25566
    <anonymous> http://localhost:1234/elm.js:25571
```

But I can explain why that one happens!

First, you are on the right track that DOM ids can cause conflicts with JS things with the same name. Watch:

<img width="438" alt="Image" src="https://github.com/user-attachments/assets/66695e76-205b-4859-91b7-7ec837ae4894" />

The DOM has a feature where if there’s an element on the page with an `id` attribute which is also a valid JS identifier, a global variable with that name is created, pointing to the element.

In your SSCCE, there’s `<div class="s e wf" id="Elm"></div>` in the HTML, which then creates a global variable called `Elm` (a.k.a. `window.Elm`) which is a reference to that div DOM node.

[The script tag for the Elm code](https://github.com/dillonkearns/elm-pages/blob/65e982194f15a1bee8474913e59e2efad72c7a51/generator/src/pre-render-html.js#L39) has `defer`:

```html
<script defer src="/elm.js" type="text/javascript"></script>
```

`defer` means that the script is guaranteed to run _after_ the whole HTML has been parsed and turned into DOM. So when elm.js runs, the global `Elm` variable pointing to your div already exists.

Elm supports `window.Elm` already existing when it runs. It then extends `window.Elm` with its modules. This allows you to load multiple Elm apps on the same page. This means that Elm assigns `Elm.Main` in your case, adding a `Main` property to the div:

<img width="327" alt="Image" src="https://github.com/user-attachments/assets/77333d9b-e329-478e-ab90-6c403235008f" />

Now on to `findPublicModules` where the error occurred for me. It comes from [elm-hot](https://github.com/klazuka/elm-hot/blob/fb2dc49e9b4fa53b51fa6088a1ac7ffa0b72557a/resources/hmr.js#L128-L143). That code tries to go through everything in `Elm` to find all the `init` functions. (In your case, it only needs to find `Elm.Main.init` but it can’t know the name(s) of your Elm modules.) It expects `window.Elm` to be a plain JS object, with no other properties than Elm modules. So it tries to do `for (var key in parent) {` where `parent` is set to `window.Elm`, but that also loops through the `align` property of the div, but `window.Elm.align` is a string, not an Elm module, so later code then fails.

My guess is that something similar happened when you got the `errorJson` error. Having `id="Elm"` seems to be asking for trouble – choose any other name!

If elm-pages _really_ want to solve this, it should be possible. Either by not relying on `window.Elm` (use a non-global instead), or by going through all the code (including elm-hot) and make sure that it doesn’t expect `window.Elm` to be “pristine”. But is it worth the trouble?

**Comment by MartinSStewart** [comment]
> choose any other name!

Yup, this is what I did.

Thanks for the explanation of what is going wrong. I don't mind having to change id names. The real issue is just that it's not at all obvious what went wrong, but given this is probably a very rare bug, maybe it's not worth doing anything about it.

**Comment by dillonkearns** [comment]
Yeah, it is unfortunate. I would consider this more of an Elm thing than an elm-pages thing. Certainly not ideal and I do try to smooth over edges as much as possible, but seems like probably a lot of effort for a pesky little thing so this one is not at the top of my personal to do list. Happy to accept a PR as always though if someone has an elegant fix for this in the future.

**Comment by dillonkearns** [comment]
Either way though, great to have this placeholder and explanation in case anyone else is scratching their heads in the future. So thank you both!
